packages/jupyter-ai/jupyter_ai/document_loaders/directory.py

```diff
@@ -120,16 +121,20 @@ def collect_filepaths(path, all_files: bool):
     if os.path.isfile(path):
         filepaths = [Path(path)]
     else:
-        filepaths = []
-        for dir, subdirs, filenames in os.walk(path):
-            # Filter out hidden filenames, hidden directories, and excluded directories,
-            # unless "all files" are requested
-            if not all_files:
-                subdirs[:] = [
-                    d for d in subdirs if not (d[0] == "." or d in EXCLUDE_DIRS)
-                ]
-                filenames = [f for f in filenames if not f[0] == "."]
-            filepaths.extend([Path(dir) / filename for filename in filenames])
+        filepaths = set()
+        for glob_path in iglob(str(path), include_hidden=all_files, recursive=True):
+            if os.path.isfile(glob_path):
+                filepaths.add(Path(glob_path))
+                continue
+            for dir, subdirs, filenames in os.walk(glob_path):
+                # Filter out hidden filenames, hidden directories, and excluded directories,
+                # unless "all files" are requested
+                if not all_files:
+                    subdirs[:] = [
+                        d for d in subdirs if not (d[0] == "." or d in EXCLUDE_DIRS)
+                    ]
+                    filenames = [f for f in filenames if not f[0] == "."]
+                filepaths.update([Path(dir) / filename for filename in filenames])
```

Comment:
There are three distinct possibilities:

1. Path refers to a specific file
2. Path refers to a directory
3. Path is a glob

Suppose we have a directory `foo/` that contains 2 children `bar.txt` and `baz/`. If I run `/learn foo/*`, the current logic also indexes the files under `foo/baz/` since `baz/` is not a file. However, this should only occur when I run `/learn foo/**/*`. We don't need to call `os.walk()` in the third case, because `iglob()` should already return a list of file paths under `foo/` recursively if a glob like `foo/**/*` is given.

I think this block should be refactored to the below for correctness:

```py
if os.path.isfile(path):
    filepaths = [Path(path)]
elif os.path.isdir(path):
    ... # existing logic
else:
    for glob_path in iglob(str(path), include_hidden=all_files, recursive=True):
        if os.path.isfile(glob_path):
            filepaths.add(Path(glob_path))
```

Reply:
This change would also allow for directories to be indexed non-recursively, that is:

- `/learn foo/`: learns everything under `foo/` recursively
- `/learn foo/**/*`: same as above
- `/learn foo/*`: **only learns files which are direct children under `foo/`**

Reply:
That's a good point. The reason I left the walk logic under glob is for cases where a user may want all the files under specific directories. For instance, if you have a directory with the following structure:

folder1/
  | file1
  | file2
folder2/
  | file3
  | file4
unwanted_folder/
  | file_a
  | file_b
test.py

You could use `glob("folder*/")` to get `["file1", "file2", "file3", "file4"]`. I am going off the assumption that if the glob pattern returns a directory, it is safe to assume that it was intended to collect items under the directory. That does leave some instances where if you walk returned directories, you may end up waking the same nested directory multiple times.

What if we only walk directories returned from glob if all the returned paths are directories? Maybe something like:

```python
def walk_directory(dir_path):
    ...

def collect_filespaths(path):
    if os.path.isfile(path):
        filepaths = [Path(path)]
    elif os.path.isdir(path):
        filepaths = walk_directory(path)
    else:
        glob_paths = glob(str(path), ...)
        filepaths = set()
        if all(map(lambda p: os.path.isdir(p), glob_paths)):
            for dir in glob_paths:
                filepaths.update(walk_directory(dir))
        else:
            for p in glob_paths:
                if os.path.isfile(glob_path):
                    filepaths.add(Path(glob_path))
    ...
```  

I think this would cover both cases. I will also have to move the extension checking out of the walk logic and move it to after paths are collected. What do you think?

Reply:
I pushed up a commit that handles this, though I can roll it back if we don't want to handle the glob directories situation